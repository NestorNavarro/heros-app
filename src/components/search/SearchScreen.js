import React, { useMemo } from 'react';
import queryString from 'query-string';
import useForm from '../../hooks/useForm';
import { useLocation } from 'react-router';
import { HeroCard } from '../hero/HeroCard';
import { getHeroesByName } from '../../selectors/getHeroesByName';

export const SearchScreen = ({ history }) => {
    const location = useLocation();

    const { q = '' } = queryString.parse(location.search);
    const [ fromValue , handleInputChange ] = useForm({
        search: q,
    });

    const { search } = fromValue;
    const heroesFiltered  = useMemo(() => getHeroesByName(q), [q]);
    const handleSearch = (e) => {
        e.preventDefault();
        history.push(`?q=${ search }`);
    }
    return (
        <div>
            <h1>Search Screen</h1>
            <hr/>
            <div className="row">
                <div className="col-5">
                    <h4>Search From</h4>
                    <hr/>
                    <form onSubmit={ handleSearch}>
                        <input
                            type="text"
                            placeholder="Find your hero"
                            className="form-control"
                            autoComplete="off"
                            name="search"
                            value={ search }
                            onChange={ handleInputChange }

                        />
                        <button
                            type="submit"
                            className="btn m-1 btn-block btn-primary"
                        >
                            Search
                        </button>
                    </form>
                </div>
                <div className="col-7">
                    <h4>Results</h4>
                    <hr/>
                    {
                        (q === '') &&
                        <div className="alert alert-info">
                            Search a hero!
                        </div>
                    }
                    {
                        (q !== '' && heroesFiltered.length === 0) &&
                        <div className="alert alert-danger">
                            Hero not Found!
                        </div>
                    }

                    {
                        heroesFiltered.map( hero => (
                            <HeroCard 
                                key={ hero.id }
                                hero={ hero }
                            />
                        ))
                    }
                </div>
            </div>
        </div>
    );
}
