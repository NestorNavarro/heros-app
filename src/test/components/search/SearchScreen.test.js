import React from 'react';
import { mount } from 'enzyme';
import { AuthContext } from '../../../auth/AuthContext';
import { SearchScreen } from '../../../components/search/SearchScreen';
import { MemoryRouter, Route } from 'react-router-dom';

describe('Test on <SearchScreen/>', () => {

    test('should show correctly with default values', () => {
        const wrapper = mount(
            <MemoryRouter initialEntries={['/search']}>
                <Route paht='/search' component={ SearchScreen }/>
            </MemoryRouter>    
        );    

        expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.alert-info').text().trim()).toBe('Search a hero!');
    }); 

    test('should show a hero correctly', () => {
        const hero = 'batman';
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
                <Route paht='/search' component={ SearchScreen }/>
            </MemoryRouter>    
        );
        expect(wrapper.find('input').prop('value')).toBe(hero);
        expect(wrapper).toMatchSnapshot();
    });

    test('should throw an hero error alert if the hero dose not exist', () => {
        const hero = 'I am not exist';
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
                <Route paht='/search' component={ SearchScreen }/>
            </MemoryRouter>    
        );
        expect(wrapper.find('input').prop('value')).toBe(hero);
        expect(wrapper.find('.alert-danger').text().trim()).toBe('Hero not Found!');
    });
    
    test('should call the history push', () => {
        const history = {
            push: jest.fn(),
        };
        const hero = 'batman';
        const wrapper = mount(
            <MemoryRouter initialEntries={[`/search?q=${hero}`]}>
                <Route 
                    paht='/search' 
                    component={ () => <SearchScreen history={history} /> }
                />
            </MemoryRouter>    
        );

        wrapper.find('input').simulate('change', {
            target: {
                name: 'search-text',
                value: hero,
            }
        });
        wrapper.find('form').prop('onSubmit')({
            preventDefault(){}
        });
        expect(history.push).toHaveBeenCalledWith(`?q=${hero}`);
    });
    
    
});
