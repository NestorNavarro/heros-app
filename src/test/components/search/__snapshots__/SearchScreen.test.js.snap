// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test on <SearchScreen/> should show a hero correctly 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search From
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value="batman"
        />
        <button
          className="btn m-1 btn-block btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="card ms-3"
        style={
          Object {
            "maxWidth": 540,
          }
        }
      >
        <div
          className="row no-gutters"
        >
          <div
            className="col-md-4"
          >
            <img
              alt="Batman"
              className="card-img"
              src="./assets/heroes/dc-batman.jpg"
            />
          </div>
          <div
            className="col-md-8"
          >
            <div
              className="card-body"
            >
              <h5
                className="card-title"
              >
                 
                Batman
              </h5>
              <p
                className="card-text"
              >
                 
                Bruce Wayne
                 
              </p>
              <p
                className="card-text"
              >
                <small
                  className="text-muted"
                >
                   
                  Detective Comics #27
                   
                </small>
              </p>
              <a
                href="/hero/dc-batman"
                onClick={[Function]}
              >
                More...
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Test on <SearchScreen/> should show correctly with default values 1`] = `
<div>
  <h1>
    Search Screen
  </h1>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-5"
    >
      <h4>
        Search From
      </h4>
      <hr />
      <form
        onSubmit={[Function]}
      >
        <input
          autoComplete="off"
          className="form-control"
          name="search"
          onChange={[Function]}
          placeholder="Find your hero"
          type="text"
          value=""
        />
        <button
          className="btn m-1 btn-block btn-primary"
          type="submit"
        >
          Search
        </button>
      </form>
    </div>
    <div
      className="col-7"
    >
      <h4>
        Results
      </h4>
      <hr />
      <div
        className="alert alert-info"
      >
        Search a hero!
      </div>
    </div>
  </div>
</div>
`;
